vendor/zkllvm-metacraft-circuits/src/circuits_impl/weigh_justification_and_finalization.h

```diff
@@ -0,0 +1,193 @@
+#pragma once
+
+#include <array>
+#include <algorithm>
+
+#include "../utils/picosha2.h"
+#include "../circuit_utils/circuit_byte_utils.h"
+#include "../circuit_utils/ssz_utils.h"
+
+using namespace circuit_byte_utils;
+using namespace ssz_utils;
+
+namespace weigh_justification_and_finalization_ {
+
+    void assert_epoch_is_not_genesis_epoch(Epoch epoch) {
+        assert_true(epoch >= 1);
+    }
+
+    Epoch get_previous_epoch(Epoch current_epoch) {
+        return current_epoch - 1;
+    }
+
+    Slot compute_start_slot_at_epoch_in_block_roots(Epoch epoch) {
+        return (epoch * SLOTS_PER_EPOCH) % SLOTS_PER_HISTORICAL_ROOT;
+    }
+
+    void verify_epoch_start_slot_root_in_block_roots(Root beacon_state_root,
+                                                     Epoch epoch,
+                                                     Root block_root,
+                                                     MerkleProof<18>
+                                                         proof) {
+        auto start_block_roots_gindex = DEPTH18_START_BLOCK_ROOTS_GINDEX;
+        auto index_in_block_roots = compute_start_slot_at_epoch_in_block_roots(epoch);
+        auto gindex = start_block_roots_gindex + index_in_block_roots;
+        ssz_verify_proof(beacon_state_root, block_root, proof, gindex);
+    }
+
+    void verify_finalized_checkpoint(Root beacon_state_root,
+                                     CheckpointVariable finalized_checkpoint,
+                                     BeaconStateLeafProof proof) {
+        auto finalized_checkpoint_leaf = hash_tree_root(finalized_checkpoint);
+        auto gindex = BEACON_STATE_FINALIZED_CHECKPOINT_GINDEX;
+        ssz_verify_proof(beacon_state_root, finalized_checkpoint_leaf, proof, gindex);
+    }
```

Comment:
Substitute the variables with the expressions in the call to ssz_verify_proof and inline this code